File: lib/parser.js
var crypto = require('crypto');

function calcHeadSize(obj) {
    var size = 2;

    // read from buffer
    if (obj instanceof Buffer) {
        if (obj[1] & 0x80) size += 4;
        switch (obj[1] & 0x7f) {
            case 126:
                size += 2;
                break;
            case 127:
                size += 8;
        }
    } else {
        // read from object
        if (obj.length > 125 && obj.length < 0x10000) 
            size += 2;
        else if (obj.length > 0xffff) 
            size += 8;
        // add masking bytes if mask true
        if (obj.mask || obj.masking && obj.masking.length == 4) 
            size += 4;
    }

    return size;
}

function readHeadBytes(incoming, head) {
    var headSize = incoming._headSize || calcHeadSize(head);
    
    incoming.fin = !!(head[0] & 0x80);
    incoming.mask = !!(head[1] & 0x80);
    incoming.rsv1 = !!(head[0] & 0x40);
    incoming.rsv2 = !!(head[0] & 0x20);
    incoming.rsv3 = !!(head[0] & 0x10);

    incoming.opcode = head[0] & 0x0f;
    incoming.length = head[1] & 0x7f;

    switch (incoming.length) {
        case 126:
            incoming.length = head.readUInt16BE(2);
            break;
        case 127:
            incoming.length = head.readUInt32BE(6);
            incoming.extendedLength = head.readUInt32BE(2);
            break;
    }

    if (incoming.mask)
        incoming.masking = head.slice(headSize - 4, headSize);

    incoming._headSize = headSize;

    return incoming;
}

function readBodyBytes(state, chunk) {
    var index = state._index;
    var length = state.length;

    var payload = chunk.slice(0, length);

    if (state.mask) {
        var masking = state.masking;
        for (var i = 0; i < payload.length; i++)
            payload[i] = payload[i] ^ masking[(i + index) % 4]; 
    }

    state._index += payload.length;

    return payload;
}

// TODO: quit if too big
function writeHeadBytes(state) {
    var opcode = state.opcode;
    var length = state.length;
    var headSize = calcHeadSize(state);

    var headBytes = new Buffer(headSize);
    headBytes.fill(0);

    // first create the length bytes because we have to 
    // know which length we write into the second byte 
    if (length > 125 && length < 0x10000) {
        headBytes.writeUInt16BE(length, 2);
        length = 126;
    } else if (length > 0xffff) {
        headBytes.writeUInt32BE(length, 6);
        length = 127;
    }

    // create the first two bytes (TODO: rsv bits are missing)
    headBytes[0] = ((state.fin) ? 0x80 : 0x00) | opcode;
    headBytes[1] = ((state.mask) ? 0x80 : 0x00) | length;

    // append the masking to the buffer
    if (state.mask || (state.masking && state.masking.length == 4)) {
        state.mask = true;
        var masking = state.masking || crypto.randomBytes(4);
        for (var i = 0; i < 4; i++)
            headBytes[headSize - 4 + i] = masking[i];
        state.masking = masking;
    }

    return headBytes;
}

function writeBodyBytes(state, body) {
    if (!state.mask || !state.masking)
        return body;

    var length = body.length;
    var masked = new Buffer(length);

    for (var i = 0; i < length; i++)
        masked[i] = body[i] ^ state.masking[i % 4];

    return masked;
}

exports.calcHeadSize = calcHeadSize;
exports.readHeadBytes = readHeadBytes;
exports.readBodyBytes = readBodyBytes;
exports.writeHeadBytes = writeHeadBytes;
exports.writeBodyBytes = writeBodyBytes;

File: src/parser.cc
#include <v8.h>
#include <node.h>
#include <node_buffer.h>
#include <iostream>

using namespace v8;

Handle<Value> CalcHeadSize(const Arguments &args) {
    HandleScope scope;
    
    Local<Object> obj = args[0]->ToObject();

    if (!node::Buffer::HasInstance(obj)) {
        ThrowException(
                Exception::TypeError(
                    String::New("Argument must be a buffer.")));

        return scope.Close(Undefined());
    }

    if (obj->GetIndexedPropertiesExternalArrayDataLength() < 2) {
        ThrowException(
                Exception::TypeError(
                    String::New("Buffer must be at least two bytes big.")));

        return scope.Close(Undefined());
    }

    unsigned char* data = static_cast<unsigned char*>(obj->GetIndexedPropertiesExternalArrayData());

    unsigned int length = 2;

    if (data[1] & 0x80)
        length += 4;

    switch (data[1] & 0x7f) {
        case 126:
            length += 2;
            break;
        case 127:
            length += 8;
            break;
    }

    return scope.Close(Number::New(length));
}

Handle<Value> ReadHeadBytes(const Arguments &args) {
    HandleScope scope;
    
    Local<Object> state = args[0]->ToObject();
    Local<Object> head = args[1]->ToObject();

    if (!state->IsObject()) {
        ThrowException(
                Exception::TypeError(
                    String::New("Argument one must be an object.")));

        return scope.Close(Undefined());
    }

    if (!node::Buffer::HasInstance(head)) {
        ThrowException(
                Exception::TypeError(
                    String::New("Argument two must be a buffer.")));

        return scope.Close(Undefined());
    }

    if (head->GetIndexedPropertiesExternalArrayDataLength() < 2) {
        ThrowException(
                Exception::TypeError(
                    String::New("Buffer must be at least two bytes big.")));

        return scope.Close(Undefined());
    }

    unsigned char* chunk = static_cast<unsigned char*>(
            head->GetIndexedPropertiesExternalArrayData());


    bool fin = !!(chunk[0] & 0x80);
    bool rsv1 = !!(chunk[0] & 0x40);
    bool rsv2 = !!(chunk[0] & 0x20);
    bool rsv3 = !!(chunk[0] & 0x10);
    bool mask = !!(chunk[1] & 0x80);

    int offset = 2;
    int opcode = chunk[0] & 0x0f;
    unsigned long length = chunk[1] & 0x7f;
    
    switch (length) {
        case 126:
            length = chunk[3];
            length += chunk[2] << 8;
            offset += 2;
            break;
        case 127:
            length = chunk[9];
            length += chunk[8] << 8;
            length += chunk[7] << 16;
            length += chunk[6] << 24;
            offset += 8;
            break;
    }

    if (length > 0xfffffff) {
        ThrowException(
                Exception::TypeError(
                    String::New("Length bigger than UInt32BE.")));

        return scope.Close(Undefined());
    }

    Persistent<Object> maskingBuffer;

    char masking[4];
    if (mask) {
        for (int i = 0; i < 4; i++)
            masking[i] = chunk[offset + i];

        maskingBuffer = node::Buffer::New(masking, 4)->handle_;
    } else {
        maskingBuffer = node::Buffer::New(0)->handle_;
    }

    state->Set(String::New("fin"), Boolean::New(fin));
    state->Set(String::New("rsv1"), Boolean::New(rsv1));
    state->Set(String::New("rsv2"), Boolean::New(rsv2));
    state->Set(String::New("rsv3"), Boolean::New(rsv3));
    state->Set(String::New("mask"), Boolean::New(mask));
    state->Set(String::New("opcode"), Number::New(opcode));
    state->Set(String::New("length"), Number::New(length));
    state->Set(String::New("masking"), maskingBuffer);

    return scope.Close(state);
}

void init(Handle<Object> exports) {
    exports->Set(String::New("calcHeadSize"), 
            FunctionTemplate::New(CalcHeadSize)->GetFunction());

    exports->Set(String::New("readHeadBytes"),
            FunctionTemplate::New(ReadHeadBytes)->GetFunction());
}

NODE_MODULE(parser, init);
